blog/_src/posts/2017-11-16-monotonicity-types-towards-a-type-system-for-eventual-consistency.md

```diff
@@ -0,0 +1,97 @@
+
+Towards this aim, we have been designing a type system for tracking monotonicity, as an extension of the simply typed lambda calculus which adds a new function abstraction construct called the *sfun*.
+
+Our approach alllows the programmer to write a special kind of function definition, the body of which is type checked using a richer type system, one which reasons about function composition rather than application. Such a function can then be proven monotone by utilizing the fact that the composition of two monotone functions is itself monotone, and other related principles.
+
+Monotonicity is a relational property; that is, its a property involving multiple applications of the same function. Such properties are blind spot for traditional type systems, so our design requires some unusual and interesting features.
+
+To give an intuition of what our type system looks like; as mentioned earlier, we introduce a special abstraction for monotone functions called an _sfun_. Since the programmer only cares about the monotonicity of a select group of functions, a special syntax construct, the sfun, serves as a signal to the type checker. Unlike the simply typed world outside of the sfun abstraction, the body of an sfun is type checked using a refinement type system, which I call the lifted type system in which monotonicity is tracked. So in our system we have a lifted (local) type system within sfuns, and a terminal (global) type system for the typed world outside of the sfun.
+
+This terminal (global) type system is a traditional type system which is embedded as a specific kind of refinement inside of the lifted (local) type system. Applying functions is a way of projecting from the lifted (local) world into the terminal (global) world.
+
+In more detail, the "global world” outside of an sfun abstraction is viewed as a degenerate subset of the “lifted world” inside the sfun abstraction. A globally well-typed sfun application is viewed as a projection onto this degenerate subset. Inside the sfun abstraction, we track the way in which each term depends on the sfun’s arguments (for example, monotonically or antitonically), but terms originating outside of the sfun (both literal constants and occurrences of variables from the global type environment) are not affected by sfun's arguments at all.
+
+Lifted reduction, the operational semantics characterized by the lifted type system, performs a step-by-step composition of the components (sfun applications, variable occurrences, literal values, etc) of an sfun. Lifted normalization of an sfun's body corresponds to the complete composition of the sfun, resulting in the sfun itself. So when our lifted type system proves the body of an sfun monotone, we know the sfun will behave monotonically when applied under global reduction.
```

Comment:
Does lifted reduction happen (1) during type-checking, or (2) when the STLC program runs?

Reply:
Somewhat along the Ben's comment: I don't understand the phrase “operational semantics characterized by the [lifted] type system” (really, any type system) at all.

Reply:
There were definitely some self-containment issues. I added a couple of paragraphs, starting with "We’re not building lifted typing derivations just for fun. Typically, a type system comes with a soundness theorem stating that ..." Hopefully this will clarify the issues that have been raised.